File: Pandacap.LowLevel/Data/TheirContent/Weasyl/WeasylFavoriteSubmission.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations
open Pandacap.PlatformBadges

type WeasylFavoriteSubmissionUser() =
    member val Login = "" with get, set
    member val DisplayName = "" with get, set
    member val Avatar = "" with get, set

type WeasylFavoriteSubmissionImage() =
    member val Url = "" with get, set

type WeasylFavoriteSubmission() =
    [<Key>]
    member val Id = Guid.Empty with get, set

    member val Submitid = 0 with get, set
    member val Title = "" with get, set
    member val PostedBy = new WeasylFavoriteSubmissionUser() with get, set
    member val PostedAt = DateTimeOffset.MinValue with get, set
    member val Thumbnails = new ResizeArray<WeasylFavoriteSubmissionImage>() with get, set
    member val Url = "" with get, set

    member val FavoritedAt = DateTimeOffset.MinValue with get, set
    member val HiddenAt = nullDateTimeOffset with get, set

    interface IFavorite with
        member this.HiddenAt
            with get () = this.HiddenAt
             and set value = this.HiddenAt <- value

        member this.FavoritedAt = this.FavoritedAt

    interface IPost with
        member _.Badges = [{ PostPlatform.GetBadge Weasyl with Text = "weasyl.com" }]
        member this.DisplayTitle = this.Title
        member this.Id = $"{this.Id}"
        member this.LinkUrl = this.Url
        member this.ProfileUrl = $"https://www.weasyl.com/~{Uri.EscapeDataString(this.PostedBy.Login)}"
        member this.Thumbnails = [
            for thumb in this.Thumbnails do {
                new IPostThumbnail with
                    member _.AltText = ""
                    member _.Url = thumb.Url
            }
        ]
        member this.PostedAt = this.PostedAt
        member this.Usericon = this.PostedBy.Avatar
        member this.Username = this.PostedBy.DisplayName

File: Pandacap.Database/WeasylFavoriteSubmission.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;

namespace Pandacap.Database
{
    public class WeasylFavoriteSubmission : IFavorite
    {
        public Guid Id { get; set; }

        public int Submitid { get; set; }

        public string Title { get; set; } = "";

        public class User
        {
            public string Login { get; set; } = "";
            public string DisplayName { get; set; } = "";
            public string? Avatar { get; set; }
        }

        public User PostedBy { get; set; } = new();

        public DateTimeOffset PostedAt { get; set; }

        public class Image : IPostThumbnail
        {
            public string Url { get; set; } = "";

            string IPostThumbnail.AltText => "";
        }

        public List<Image> Thumbnails { get; set; } = [];

        public string ProfileUrl { get; set; } = "";

        public string Url { get; set; } = "";

        public DateTimeOffset FavoritedAt { get; set; }

        public DateTimeOffset? HiddenAt { get; set; }

        public Badge Badge => Badges.Weasyl.WithHostFromUriString(Url);

        public string DisplayTitle => Title;

        string IPost.Id => $"{Id}";

        public string InternalUrl => Url;

        public string ExternalUrl => Url;

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Thumbnails;

        public string? Username => PostedBy.DisplayName;

        public string? Usericon => PostedBy.Avatar;
    }
}
